ndb/src/google/cloud/ndb/model.py

```diff
@@ -336,6 +339,124 @@ def __hash__(self):
+    user <-> base conversions: the user -> base conversion goes from more
+    sophisticated to less sophisticated, while the base -> user conversion goes
+    from less sophisticated to more sophisticated. For example, see the
+    relationship between :class:`BlobProperty`, :class:`TextProperty` and
+    :class:`StringProperty`.
+
+    In addition to ``_to_base_type()`` and ``_from_base_type()``, the
+    ``_validate()`` method is also a composable API.
+
+    The validation API distinguishes between "lax" and "strict" user values.
+    The set of lax values is a superset of the set of strict values. The
+    ``_validate()`` method takes a lax value and if necessary converts it to
+    a strict value. This means that when setting the property value, lax values
+    are accepted, while when getting the property value, only strict values
+    will be returned. If no conversion is needed, ``_validate()`` may return
+    :data`None`. If the argument is outside the set of accepted lax values,
+    ``_validate()`` should raise an exception, preferably :exc:`TypeError` or
+    :exc:`.BadValueError`.
+
+    Example/boilerplate:
+
+    .. code-block:: python
+
+        def _validate(self, value):
+            # Lax user value to strict user value.
+            if not isinstance(value, <top type>):
+                raise TypeError(...)  # Or datastore_errors.BadValueError(...).
+
+        def _to_base_type(self, value):
+            # (Strict) user value to base value.
+            if isinstance(value, SomeUserType):
+                return SomeBaseType(value)
+
+        def _from_base_type(self, value):
+            # Base value to (strict) user value.'
+            if not isinstance(value, SomeBaseType):
+                return SomeUserType(value)
+
+    Things that ``_validate()``, ``_to_base_type()`` and
+    ``_from_base_type()`` do **not** need to handle:
+
+    * :data:`None`: They will not be called with :data:`None` (and if they
+      return :data:`None`, this means that the value does not need conversion).
+    * Repeated values: The infrastructure (:meth:`_get_user_value` and
+      :meth:`_get_base_value`) takes care of calling ``_from_base_type()`` or
+      ``_to_base_type()`` for each list item in a repeated value.
+    * Wrapping values in :class:`_BaseValue`: The wrapping and unwrapping is
+      taken care of by the infrastructure that calls the composable APIs.
+    * Comparisons: The comparison operations call ``_to_base_type()`` on
+      their operand.
+    * Distinguishing between user and base values: the infrastructure
+      guarantees that ``_from_base_type()`` will be called with an
+      (unwrapped) base value, and that ``_to_base_type()`` will be called
+      with a user value.
+    * Returning the original value: if any of these return :data:`None`, the
+      original value is kept. (Returning a different value not equal to
+      :data:`None` will substitute the different value.)
+
```

Comment:
While clear the audience needs to know about `_validate`, `_to_base_type`, and `_from_base_type`, it's far less clear the audience needs to know about `_get_user_value`, `_get_base_value`, or `_BaseValue`. See question above about audience.

Reply:
Jinx!

Reply:
:+1: